src/common/hashes/shamap.ts

```diff
@@ -0,0 +1,263 @@
+import {HASH_INNER_NODE, HASH_INNER_NODE_V2, HASH_TX_ID, HASH_TX_NODE, HASH_LEAF_NODE} from './hashprefix'
+import hash from './sha512hash'
+const HEX_ZERO = '0000000000000000000000000000000000000000000000000000000000000000'
+export const TYPE_INNER = 1
+export const TYPE_TRANSACTION_NM = 2
+export const TYPE_TRANSACTION_MD = 3
+export const TYPE_ACCOUNT_STATE = 4
+
+/**
+ * Abstract class representing a node in a SHAMap tree.
+ *
+ * Can be either SHAMapTreeNodeInner, SHAMapTreeNodeInnerV2
+ * or SHAMapTreeNodeLeaf.
+ *
+ * @class
+ */
+export abstract class Node {
+  public constructor() {}
+  public add_item(_tag: string, _node: Node): void {
+    throw new Error('Called unimplemented virtual method SHAMapTreeNode#add_item.')
+  }
+  public get hash(): string|void {
+    throw new Error('Called unimplemented virtual method SHAMapTreeNode#hash.');
+  }
+}
+
+/**
+ * Inner (non-leaf) node in a SHAMap tree.
+ * @param {Number} depth (i.e. how many parent inner nodes)
+ * @class
+ */
+export class InnerNode extends Node {
+  public leaves: object
```

Comment:
IIUC this is an object with keys that are characters 0-F? 

You could get better typing here if you change from 0-15 and then enforce that constraint I think?

Reply:
I'm not sure its a good idea to change the behavior of this but I did so in my last commit lmk what you think https://github.com/ripple/ripple-lib/pull/1039/commits/f63a364cd3e7668bbc6958b8a3051f136588b29e